src/backend/processor/diagnostics.ts

```diff
@@ -0,0 +1,196 @@
+import { Event, EventEmitter, ExtensionContext, TextEditor, Uri, window } from 'vscode';
+import { parseDiagnostics } from '../parser';
+import { CheckerMetadata, DiagnosticEntry, DiagnosticFile } from '../types';
+import { ExtensionApi } from '../api';
+
+/**
+ * API interface that provides Diagnostics data.  
+ * Access the active instance via ExtensionApi.
+ */
+export class DiagnosticsApi {
+    private _openedFiles: string[] = [];
+
+    /** 
+     * Content based on the contents of loaded .plist files.
+     * Key: Source code file, value: .plist analysis files
+     */
+    private _diagnosticSourceFiles: Map<string, string[]> = new Map();
+
+    /**
+     * Loaded diagnostic files.
+     * Key: .plist analysis file, value: file's contents
+     */
+    private _diagnosticEntries: Map<string, DiagnosticFile> = new Map();
+
+    private _selectedEntry?: {
+        file: string,
+        idx: number
+    };
+    public get selectedEntry(): {
+        position: {readonly file: string, readonly idx: number},
+        diagnostic: DiagnosticEntry
+    } | undefined {
+        if (this._selectedEntry) {
+            const activeFile = this._diagnosticEntries.get(this._selectedEntry.file);
+            const diagnostic = activeFile?.diagnostics[this._selectedEntry.idx];
+
+            if (diagnostic) {
+                return {
+                    position: this._selectedEntry!,
+                    diagnostic
+                };
+            }
+        }
+
+        return undefined;
+    }
+
+    public setActiveReport(position: {file: string, idx: number} | undefined) {
+        if (position) {
+            const activeFile = this._diagnosticEntries.get(position.file);
+            const diagnostic = activeFile?.diagnostics[position.idx];
+
+            if (diagnostic) {
+                this._selectedEntry = position;
+                return;
+            }
+        }
+
+        this._selectedEntry = undefined;
+    }
+
+    constructor(ctx: ExtensionContext) {
+        ctx.subscriptions.push(this._diagnosticsUpdated = new EventEmitter());
+        window.onDidChangeVisibleTextEditors(this.onDocumentsChanged, this, ctx.subscriptions);
+        ExtensionApi.metadata.metadataUpdated(this.onMetadataUpdated, this, ctx.subscriptions);
+
+        this.init();
+    }
+
+    private init() {
+        // Initialize the currently opened windows
+        this.onDocumentsChanged(window.visibleTextEditors);
+    }
+
+    /** Prefer calling ``reloadDiagnosticsAsync`` in an async context */
+    reloadDiagnostics(forceReload?: boolean) {
+        this.reloadDiagnosticsAsync(forceReload)
```

Comment:
Do not we need to use `await` here? So this function will return with a promise too. Because in this case if I do something right after the reloadDiagnostics function call, it will not wait while the reloadDiagnostics fully finishes:
```js
reloadDiagnostics();
console.log("It will be executed first.");
```
With the await we will be able to do this:
```js
reloadDiagnostics().then(() => {
  console.log("It will be executed first.");
});
``

Reply:
You're right, the concept of the function was useless, as an user could ``reloadDiagnosticsAsync()`` instead in a non-async environment.

Replaced it with a ``reloadDiagnosticsSafe()``, which handles the rejection of ``reloadDiagnostics`` with an error popup to the user. (Is there a naming convention for this type of functions? I'm not sure about the name)